Source/WebCore/Modules/cookie-store/CookieStore.cpp

```diff
@@ -392,18 +558,43 @@ ScriptExecutionContext* CookieStore::scriptExecutionContext() const
 
 void CookieStore::eventListenersDidChange()
 {
+    // FIXME: This should work for service worker contexts as well.
+    if (!is<Document>(scriptExecutionContext()))
+        return;
+
     bool hadChangeEventListener = m_hasChangeEventListener;
     m_hasChangeEventListener = hasEventListeners(eventNames().changeEvent);
 
-    if (hadChangeEventListener == m_hasChangeEventListener || !m_cookieJar)
+    if (hadChangeEventListener == m_hasChangeEventListener)
+        return;
+
+    auto* document = downcast<Document>(scriptExecutionContext());
+    if (!document)
+        return;
+
+    auto* page = document->page();
+    if (!page)
         return;
 
 #if HAVE(COOKIE_CHANGE_LISTENER_API)
+    auto& cookieJar = page->cookieJar();
+    auto host = document->url().host().toString();
     if (m_hasChangeEventListener)
-        m_cookieJar->addChangeListener(m_host, *this);
+        cookieJar.addChangeListener(host, *this);
     else
-        m_cookieJar->removeChangeListener(m_host, *this);
+        cookieJar.removeChangeListener(host, *this);
 #endif
 }
 
+RefPtr<DeferredPromise> CookieStore::takePromise(uint64_t promiseIdentifier)
+{
+    auto it = m_promises.find(promiseIdentifier);
```

Comment:
This should be as easy as `return m_promises.take(promiseIdentifier);`.

If this doesn't work, then use a `RefPtr<>` instead of a `Ref<>` for the value in your HashMap to make it work.